w-dms/sqliteDriver.go

```diff
@@ -0,0 +1,279 @@
+	}
+	typeID, err := d.findTypeID(t.Type)
+	if err != nil {
+		return err
+	}
+	res, err := transaction.Exec(updateClause, t.Alias, t.Description, typeID, t.Timestamp, t.EstimateTime, t.RealTime, t.ID)
+	if err != nil {
+		return err
+	}
+	count, err := res.RowsAffected()
+	if err == sql.ErrNoRows || count == 0 {
+		return errTaskNotFound
+	} else if err != nil {
+		return err
+	}
+	if err = addTaskTagsAndRemindersInTransaction(t, transaction); err != nil {
+		return err
+	}
+	canCommit = true
+	return nil
+}
+
+func (d *sqliteDriver) Delete(t Task) error {
+	id, err := strconv.ParseInt(t.ID.(string), 10, 0)
+	if err != nil {
+		return err
+	}
+	t.ID = id
+	var canCommit = false
+	transaction, err := d.db.Begin()
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if canCommit {
+			transaction.Commit()
+		} else {
+			transaction.Rollback()
+		}
+	}()
+	if err = cleanTagsAndRemindersInTransaction(t, transaction); err != nil {
+		return err
+	}
+	res, err := transaction.Exec("delete from Tasks where id = ?", t.ID)
+	if err != nil {
+		return err
+	}
+	count, err := res.RowsAffected()
+	if err == sql.ErrNoRows || count == 0 {
+		return errTaskNotFound
+	} else if err != nil {
+		return err
+	}
+	canCommit = true
+	return nil
+}
+
+func (d *sqliteDriver) findTypeID(typeName string) (int, error) {
+	row := d.db.QueryRow("select id from Types where type = ?", typeName)
+	var id int
+	err := row.Scan(&id)
+	if err != nil {
+		return 0, err
+	}
+	return id, nil
+}
+
+func addTaskTagsAndRemindersInTransaction(t Task, tx *sql.Tx) error {
+	var tagStm, remStm *sql.Stmt
+	defer func() {
+		if tagStm != nil {
+			tagStm.Close()
+		}
+		if remStm != nil {
+			remStm.Close()
+		}
+	}()
+	tagStm, err := tx.Prepare("insert into Tags(task_id, tag) values(?, ?)")
+	if err != nil {
+		return err
+	}
+	for _, tag := range t.Tags {
+		_, err = tagStm.Exec(t.ID, tag)
+		if err != nil {
+			return err
+		}
+	}
+	remStm, err = tx.Prepare("insert into Reminders(task_id, reminder) values(?, ?)")
+	if err != nil {
+		return err
+	}
+	for _, reminder := range t.Reminders {
+		_, err = remStm.Exec(t.ID, reminder)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func cleanTagsAndRemindersInTransaction(task Task, tx *sql.Tx) error {
+	_, err := tx.Exec("delete from Tags where task_id = ?", task.ID)
+	if err != nil {
+		return err
+	}
+	_, err = tx.Exec("delete from Reminders where task_id = ?", task.ID)
+	return err
+}
+
+func (d *sqliteDriver) readByCondition(condition string, args ...interface{}) (TaskList, error) {
+	var taskMap = make(map[int]*Task, 0)
+	var rows *sql.Rows
+	var err error
+	rows, err = d.db.Query(selectCaluse+" "+condition, args...)
+	if err != nil {
+		return nil, err
+	}
+	var id int
+	var ts int64
+	var alias, typ string
+	var descr, rt, et, tag, rem sql.NullString
+	for rows.Next() {
+		err := rows.Scan(&id, &alias, &descr, &ts, &rt, &et, &typ, &rem, &tag)
+		if err != nil {
+			return nil, err
+		}
+		task, exists := taskMap[id]
```

Comment:
Task has two nested sets. If map doesn't contain task, then I need to create new one and put it into the map. If the task was already created, then I need to push values into it. I don't know how to avoid this check...